File: dart/stash/lib/src/safe.dart
import 'dart:convert';
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'comm.dart';
import 'database.dart';
import 'db.dart';
import 'filesystem.dart';
import 'identity.dart';
import 'loader.dart';

class Config {
  int quota = 0;
  String description = '';

  fromJson(Map<String, dynamic> json) {
    quota = json['quota'];
    description = json['description'];
  }

  toJson() {
    return {
      'quota': quota,
      'description': description,
    };
  }
}

final emptyConfig = Config();

typedef Groups = Map<String, Set<String>>;

class Safe {
  int hnd = 0;

  static int grant = 0;
  static int revoke = 1;
  static int curse = 2;
  static int endorse = 3;

  Safe.create(DB db, Identity identity, String url, {Config? config}) {
    config ??= emptyConfig;
    var fun = stashLibrary!.lookupFunction<ArgsISSS, ArgsiSSS>('stash_createSafe');
    var res = fun(db.hnd, jsonEncode(identity.toJson()).toNativeUtf8(),
        url.toNativeUtf8(), jsonEncode(config.toJson()).toNativeUtf8());
    hnd = res.handle;
  }

  Safe.open(DB db, Identity identity, String url) {
    var fun = stashLibrary!.lookupFunction<ArgsISS, ArgsiSS>('stash_openSafe');
    var res = fun(db.hnd, jsonEncode(identity.toJson()).toNativeUtf8(),
        url.toNativeUtf8());
    hnd = res.handle;
  }

  void close() {
    var fun = stashLibrary!.lookupFunction<ArgsI, Argsi>('stash_closeSafe');
    fun(hnd);
  }

  Groups updateGroup(String groupName, int change, List<String> users) {
    var fun =
        stashLibrary!.lookupFunction<ArgsISIS, ArgsiSiS>('stash_updateGroup');
    var res = fun(hnd, groupName.toNativeUtf8(), change,
        jsonEncode(users).toNativeUtf8());
    return res.map
        .map((k, v) => MapEntry(k, (v as Map<String, dynamic>).keys.toSet()));
  }

  Groups getGroups() {
    var fun = stashLibrary!.lookupFunction<ArgsI, Argsi>('stash_getGroups');
    var res = fun(hnd);
    return res
        .map
        .map((k, v) => MapEntry(k, (v as Map<String, dynamic>).keys.toSet()));
  }

  List<String> getKeys(String groupName) {
    var fun = stashLibrary!.lookupFunction<ArgsIS, ArgsiS>('stash_getKeys');
    var res = fun(hnd, groupName.toNativeUtf8());
    return List<String>.from(res.list);
  }

  Filesystem openFS() {
    var fun = stashLibrary!.lookupFunction<ArgsI, Argsi>('stash_openFS');
    var res = fun(hnd);
    return Filesystem(res.handle);
  }

  Database openDatabase(String groupName, Map<String, String> ddls) {
    var fun = stashLibrary!.lookupFunction<ArgsISS, ArgsiSS>('stash_openDatabase');
    var s = jsonEncode(ddls);
    var res = fun(hnd, groupName.toNativeUtf8(), s.toNativeUtf8());
    return Database(res.handle);
  }

  Comm openComm() {
    var fun = stashLibrary!.lookupFunction<ArgsI, Argsi>('stash_openComm');
    var res = fun(hnd);
    return Comm(res.handle);
  }
}

File: java/lib/src/main/java/ink/francesco/stash/Safe.java
package ink.francesco.stash;

import java.util.Set;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;




public class Safe {

    public static String usrGroup = "usr";
    public static String adminGroup = "admin";

    static public class Config {
    @JsonProperty
    long quota;
    @JsonProperty
    String description;

    String toJson() throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        return mapper.writeValueAsString(this);
    }
}

    static public enum GroupChange {
        Add(1), Remove(2);
        
        private final int value;
        GroupChange(int value) {
            this.value = value;
        }
    }

    long hnd;
    ObjectMapper mapper = new ObjectMapper();

    static public Safe create(DB db, Identity identity, String url, 
                Config config ) throws Exception {
        Result r = StashLibrary.instance.stash_createSafe(db.hnd, identity.toJson(), url, config.toJson());
        r.check();
        r.check();
        return new Safe(r.hnd);
    }

    static public Safe open(DB db, Identity identity, String url) throws Exception {
        Result r = StashLibrary.instance.stash_openSafe(db.hnd, identity.toJson(), url);
        r.check();
        return new Safe(r.hnd);
    }

    public void close()  {
        StashLibrary.instance.stash_closeSafe(hnd);
    }

    private Safe(long hnd) {
        this.hnd = hnd;
    }

    public Set<String> updateGroup(String groupName, GroupChange change, String[] users) throws Exception {
        Result r = StashLibrary.instance.stash_updateGroup(hnd, groupName, change.value, mapper.writeValueAsString(users));
        r.check();
        return r.map(String.class, Boolean.class).keySet();
    }

    public Set<String> getGroups() throws Exception {
        Result r = StashLibrary.instance.stash_getGroups(hnd);
        r.check();
        return r.map(String.class, Boolean.class).keySet();
    }

    public Set<String> getKeys(String groupName, long expectedMinimumLenght) throws Exception {
        Result r = StashLibrary.instance.stash_getKeys(hnd, groupName, expectedMinimumLenght);
        r.check();
        return r.map(String.class, Boolean.class).keySet();
    }

    public FS openFS() throws Exception {
        Result r = StashLibrary.instance.stash_openFS(hnd);
        r.check();
        return new FS(r.hnd);
    }

    public Database openDatabase(String groupName, Map<String,String> ddls) throws Exception {
        Result r = StashLibrary.instance.stash_openDatabase(hnd, groupName, mapper.writeValueAsString(ddls));
        r.check();
        return new Database(r.hnd);
    }

    public Comm openComm() throws Exception {
        Result r = StashLibrary.instance.stash_openComm(hnd);
        r.check();
        return new Comm(r.hnd);
    }
}
